presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1104,34 +1107,32 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
```

Comment:
What you write makes sense. Though if, I understand correctly changing the code the way you suggest, is total rebuild of how join is planned today. We would need to add original join condition into `JoinNode` (it is not there right now - only list of `EquiJoinClause`s). And then move most of the logic from `RelationPlanner.visitJoin` to one or more optimizers.

As currently all the detailed planning happens in `RelationPlanner.visitJoin` I cannot move the normalization into `SimplifyExpressions` as this one is executing *after* `RelationPlanner.visitJoin`. 

I would rather not put this Join planning refactor as part of this PR. Honestly I am not sure if with my current knowledge I would be able to efficiently implement this at all :)